File: pipeline/build_worklist.js
// Phase 3a: build a summarization worklist for a topic (papers with a PDF, not yet summarized).
// Usage: node --experimental-sqlite pipeline/build_worklist.js <topicId>
'use strict';
const fs = require('node:fs');
const path = require('node:path');
const { open, ROOT } = require('./lib/db');

const fileId = (id) => id.replace(/[^a-z0-9._-]/gi, '_').slice(0, 180);

function main() {
  const topicId = process.argv[2];
  if (!topicId) { console.error('usage: build_worklist.js <topicId>'); process.exit(1); }
  const db = open();
  const rows = db.prepare(
    `SELECT p.*, pt.relevance, pt.relevance_reason
       FROM papers p JOIN paper_topic pt ON pt.paper_id=p.id
      WHERE pt.topic_id=? AND p.status='pdf_downloaded'
      ORDER BY pt.rank`).all(topicId);
  db.close();

  const work = rows.map((r) => {
    const sdir = path.join(ROOT, 'store', 'summaries', r.slug || fileId(r.id));
    return {
      id: r.id,
      title: r.title,
      authors: JSON.parse(r.authors || '[]'),
      year: r.year,
      venue: r.venue,
      citation_count: r.citation_count,
      doi: r.doi,
      landing_url: r.landing_url,
      is_edge: !!r.is_edge,
      relevance: r.relevance,
      relevance_reason: r.relevance_reason,
      text_path: r.text_path ? path.join(ROOT, r.text_path) : null,
      pdf_path: r.pdf_path ? path.join(ROOT, r.pdf_path) : null,
      summary_dir: sdir,
      summary_path: path.join(sdir, 'v1.md'),
    };
  });

  const out = path.join(ROOT, 'topics', topicId, 'summarize_worklist.json');
  fs.writeFileSync(out, JSON.stringify({ topicId, total: work.length, work }, null, 2));
  console.log(`worklist: ${work.length} papers -> topics/${topicId}/summarize_worklist.json`);
  console.log(`  with-text: ${work.filter((w) => w.text_path).length}  pdf-only: ${work.filter((w) => !w.text_path && w.pdf_path).length}`);
}
main();

File: pipeline/build_worklist.py
"""Build a summarization worklist for a topic (papers with a PDF, not yet summarized).
Usage: python3 pipeline/build_worklist.py <topicId>
"""
import json
import re
import sys

from lib.db import open_db, ROOT


def file_id(pid):
    return re.sub(r"[^a-z0-9._-]", "_", pid, flags=re.I)[:180]


def main():
    if len(sys.argv) < 2:
        print("usage: build_worklist.py <topicId>", file=sys.stderr)
        sys.exit(1)
    topic_id = sys.argv[1]
    conn = open_db()
    rows = conn.execute(
        """SELECT p.*, pt.relevance, pt.relevance_reason
             FROM papers p JOIN paper_topic pt ON pt.paper_id=p.id
            WHERE pt.topic_id=? AND p.status='pdf_downloaded'
            ORDER BY pt.rank""", (topic_id,)).fetchall()
    conn.close()

    work = []
    for r in rows:
        base = r["slug"] or file_id(r["id"])
        sdir = ROOT / "store" / "summaries" / base
        work.append({
            "id": r["id"], "title": r["title"], "authors": json.loads(r["authors"] or "[]"),
            "year": r["year"], "venue": r["venue"], "citation_count": r["citation_count"],
            "doi": r["doi"], "landing_url": r["landing_url"], "is_edge": bool(r["is_edge"]),
            "relevance": r["relevance"], "relevance_reason": r["relevance_reason"],
            "quality_tier": r["quality_tier"], "quality_signals": r["quality_signals"],
            "text_path": str(ROOT / r["text_path"]) if r["text_path"] else None,
            "pdf_path": str(ROOT / r["pdf_path"]) if r["pdf_path"] else None,
            "summary_dir": str(sdir),
            "summary_path": str(sdir / "v1.md"),
        })

    out = ROOT / "topics" / topic_id / "summarize_worklist.json"
    out.write_text(json.dumps({"topicId": topic_id, "total": len(work), "work": work},
                              ensure_ascii=False, indent=2), encoding="utf-8")
    with_text = sum(1 for w in work if w["text_path"])
    print(f"worklist: {len(work)} papers -> topics/{topic_id}/summarize_worklist.json")
    print(f"  with-text: {with_text}  pdf-only: {sum(1 for w in work if not w['text_path'] and w['pdf_path'])}")


if __name__ == "__main__":
    main()
